torch_xla/csrc/ops/device_data.h

```diff
@@ -10,10 +10,19 @@ class DeviceData : public XlaNode {
  public:
   DeviceData(std::shared_ptr<torch::lazy::BackendData> data);
 
+  DeviceData(std::shared_ptr<torch::lazy::BackendData> data,
+             torch::lazy::OpList ops, xla::Shape xla_shape,
+             xla::OpSharding sharding);
+
   std::string ToString() const override;
 
+  torch::lazy::NodePtr Clone() const override;
+
   torch::lazy::NodePtr Clone(torch::lazy::OpList operands) const override;
 
+  torch::lazy::NodePtr CloneWithSharding(
```

Comment:
What happens to the other node types? We can attach sharding to any.

Reply:
Unless they have this method implemented it will crash at runtime. So far, only generic and device data are effected. If we need to support sharding for any node, we will need to add these functions to each node class.

Reply:
I see.. I think we have data & linear model sharding test cases. Could you try with conv layer as well, to see if that would also require touching other nodes? I hope we can cover most of common cases generic & data.